sbom/spdx/encoder.go

```diff
@@ -0,0 +1,444 @@
+}
+
+func WithDocumentComment(comment string) Option {
+	return func(e *Encoder) {
+		e.DocumentComment = comment
+	}
+}
+
+// Encode encodes a [claircore.IndexReport] that writes to w.
+// We first convert the IndexReport to an SPDX doc of the latest version, then
+// convert that doc to the specified version. We assume there's no data munging
+// going from latest to the specified version.
+func (e *Encoder) Encode(ctx context.Context, w io.Writer, ir *claircore.IndexReport) error {
+	spdx, err := e.parseIndexReport(ctx, ir)
+	if err != nil {
+		return err
+	}
+
+	// TODO(blugo): support SPDX versions before 2.3
+	var tmpConverterDoc common.AnyDocument
+	switch e.Version {
+	case V2_3:
+		// parseIndexReport currently returns a v2_3.Document so do nothing
+		tmpConverterDoc = spdx
+	default:
+		return fmt.Errorf("unknown SPDX version: %v", e.Version)
+	}
+
+	switch e.Format {
+	case JSONFormat:
+		if err := spdxjson.Write(tmpConverterDoc, w); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return fmt.Errorf("unknown requested format: %v", e.Format)
+}
+
+func (e *Encoder) parseIndexReport(ctx context.Context, ir *claircore.IndexReport) (*v2_3.Document, error) {
+	creators := make([]v2common.Creator, len(e.Creators))
+	for i, creator := range e.Creators {
+		creators[i] = v2common.Creator{
+			Creator:     creator.Creator,
+			CreatorType: creator.CreatorType,
+		}
+	}
+
+	// Initial metadata
+	out := &v2_3.Document{
+		SPDXVersion:       v2_3.Version,
+		DataLicense:       v2_3.DataLicense,
+		SPDXIdentifier:    "DOCUMENT",
+		DocumentName:      e.DocumentName,
+		DocumentNamespace: e.DocumentNamespace,
+		CreationInfo: &v2_3.CreationInfo{
+			Creators: creators,
+			Created:  time.Now().Format("2006-01-02T15:04:05Z"),
+		},
+		DocumentComment: e.DocumentComment,
+	}
+
+	pkgs := make(map[int]*v2_3.Package)
+	var pkgIds []int
+	dists := make(map[int]*v2_3.Package)
+	var distIds []int
+	repos := make(map[int]*v2_3.Package)
+	var repoIds []int
+	pkgRels := map[int][]*v2_3.Relationship{}
+	for _, r := range ir.IndexRecords() {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+
+		rPkgId, err := strconv.Atoi(r.Package.ID)
+		if err != nil {
+			return nil, err
+		}
+
+		pkg, ok := pkgs[rPkgId]
+		// Record the package if we haven't seen it yet.
+		if !ok {
+			pkgDB := ""
+			for _, env := range ir.Environments[r.Package.ID] {
+				if env.PackageDB != "" {
+					pkgDB = env.PackageDB
+					break
+				}
+			}
+
+			pkgPurpose := "APPLICATION"
+			if r.Package.Kind != claircore.BINARY {
+				pkgPurpose = "SOURCE"
+			}
+
+			pkg = newSpdxPackageFromPackage(r.Package)
+			pkg.PackageFileName = pkgDB
+			pkg.FilesAnalyzed = true
+			pkg.PrimaryPackagePurpose = pkgPurpose
+
+			pkgs[rPkgId] = pkg
+			pkgIds = append(pkgIds, rPkgId)
+
+			if r.Package.Source != nil && r.Package.Source.Name != "" {
+				rSrcPkgId, err := strconv.Atoi(r.Package.Source.ID)
+				if err != nil {
+					return nil, err
+				}
+
+				srcPkg, ok := pkgs[rSrcPkgId]
```

Comment:
won't block the PR over this, but wondering why we add the source here instead of just waiting until we see it? Do we not always see it? In the case we add it here and then see it again later, do we risk potentially duplicating data?

Reply:
@crozzy will likely be able to explain this better, but my understanding is that source packages aren't always recorded as root-level packages. In fact, I _think_ rhcc source packages are the only ones that are also recorded as root-level packages.
For clarity, when I say "root-level packages" here, I'm referring to `IndexReport.Packages` as opposed to `IndexReport.Packages.Source`.